### PR TITLE
docs(decision-record): proposal for entity timestamp

### PR DESCRIPTION
**Pull request by diegogomez-zf**
## What this PR changes/adds

Adds a Technical document explaining what, why and how the changes related to the timestamp will be implemented in the entities

## Why it does that

In order to show a proposal of the changes to be implemented for the issue #1370 

## Linked Issue(s)

#1370

## Checklist

- [x] performed checkstyle check locally?
- [x] added/updated relevant documentation?
- [x] formatted title correctly?

### CODE REVIEW

**Review by ndr-brt — APPROVED** [review]

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+```java
+public class Asset {
+
+    ...
+    private long createdTimestamp;
+    
+    ...
+
+    @NotNull
+    public Long getCreatedTimestamp() {
```
Comment (jimmarino) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:40]: Why does this return Long and not long?

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+            return this;
+        }
+    ...
+    }
+}
+```
+
+File `Schema.sql`
+
+```roomsql
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:136]:
```suggestion
```sql
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+            Objects.requireNonNull(definition.createdTimestamp);
+            return definition;
+        }
+    ...
+    }
+```
+
+File `Schema.sql`
+
+```roomsql
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:277]:
```suggestion
```sql
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+            return new ContractAgreement(id, providerAgentId, consumerAgentId, contractSigningDate, contractStartDate, contractEndDate, policy, assetId, createdTimestamp);
+        }
+        ...
+    }
+}
+```
+
+file `schema.sql`
+
+```roomsql
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:383]:
```suggestion
```sql
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+            return this;
+        }
+        ...
+
+    }
+```
+
+File `Schema.sql`
+
+```roomsql
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:59]:
```suggestion
```sql
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 6]
```diff
@@ -0,0 +1,449 @@
+# Entities timestamp
+
+## Decision
+
+An entity represents any singular, identifiable and separate business object inside an application.
+
+The decision based on the definition explained above is to add a timestamp to the EDC entities identifying their
+creation. These Entities are `ContractAgreement`, `ContractDefinition`, `PolicyDefinition`, and `Asset`.
+The `ContractNegotiation`
+and `TransferProcess` entities have it already.
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:10]:
```suggestion
The `ContractNegotiation` and `TransferProcess` entities have it already.
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+Class `ContractDefinition`
+
+```Java
+public class ContractDefinition {
+    ...
+    private long createdTimestamp;
+    ...
+
+    @NotNull
+    public Long getCreatedTimestamp() {
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:249]: Same as @jimmarino already mentioned

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+...
+CREATE TABLE IF NOT EXISTS edc_contract_agreement
+(
+    agr_id            VARCHAR NOT NULL
+        CONSTRAINT contract_agreement_pk
+            PRIMARY KEY,
+    provider_agent_id VARCHAR,
+    consumer_agent_id VARCHAR,
+    created_timestamp BIGINT,
+    signing_date      BIGINT,
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:393]: Isn't the created date the same as the signing date?

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md (hunk 8 above)
Comment (ndr-brt) [reply docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:389]: @florianrusch-zf no, they aren't the same, they have two different meanings: 
one represents the time the entity was created, the other is a domain value that represent the time the agreement was signed. Currently they have the same value but the meaning is quite different.
Moreover, `createdTimestamp` is in milliseconds, `signingDate` in epoch seconds

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 9]
```diff
@@ -349,38 +346,37 @@ public class ContractAgreement {
     /**
      * Creation timestamp of the {@link ContractAgreement}.
      *
-     * @return contract id
-     */
-    @NotNull
+     * @return timestamp when the ContractAgreement was created
+     * */
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:350]:
```suggestion
     */
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 10]
```diff
@@ -349,38 +346,37 @@ public class ContractAgreement {
...
-    ...
+...
 
     @JsonPOJOBuilder(withPrefix = "")
     public static class Builder {
 
         private String id;
         private long createdTimestamp;
-        ...
+ ...
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:361]:
```suggestion
...
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 11]
```diff
@@ -349,38 +346,37 @@ public class ContractAgreement {
...
 
         public Builder createdTimestamp(long createdTimestamp) {
             this.createdTimestamp = createdTimestamp;
             return this;
         }
-        ...
+ ...
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:367]:
```suggestion
...
```

File: docs/developer/decision-records/2022-06-28-entities-timestamp/README.md [hunk 12]
```diff
@@ -349,38 +346,37 @@ public class ContractAgreement {
...
 
         public ContractAgreement build() {
             return new ContractAgreement(id, providerAgentId, consumerAgentId, contractSigningDate, contractStartDate, contractEndDate, policy, assetId, createdTimestamp);
         }
-        ...
+ ...
```
Comment (florianrusch-zf) [inline docs/developer/decision-records/2022-06-28-entities-timestamp/README.md:372]:
```suggestion
...
```

**Review by florianrusch-zf — APPROVED** [review]

**Review by jimmarino — CHANGES_REQUESTED** [review]

**Comment by jimmarino** [comment]
> This PR neglects two key elements:
> 
> * why don't we use an `Entity` base class, as superclass of `StatefulEntity`, that contains the necessary fields.
> * the fact that entity updates are also to be tracked, at least according to the linked issue (#1370), was completely ignored.
> 
> Considering the fact that these are two key aspects of this DR, and also keeping in mind the amount of time that has already been consumed by this seemingly simple issue, my recommendation is to close and/or reassign.

I'm completely confused about this DR. Is the goal to introduce the concept of an Entity (which may include timestamps), or is it to simply add a timestamp to a set of types?

My suggestion would be to go back and clarify the goal of the DR and reformulate it more succinctly.

**Comment by ndr-brt** [comment]
> > This PR neglects two key elements:
> > 
> > * why don't we use an `Entity` base class, as superclass of `StatefulEntity`, that contains the necessary fields.
> > * the fact that entity updates are also to be tracked, at least according to the linked issue ([Core: Add timestamp to entities #1370](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/1370)), was completely ignored.
> > 
> > Considering the fact that these are two key aspects of this DR, and also keeping in mind the amount of time that has already been consumed by this seemingly simple issue, my recommendation is to close and/or reassign.
> 
> I'm completely confused about this DR. Is the goal to introduce the concept of an Entity (which may include timestamps), or is it to simply add a timestamp to a set of types?
> 
> My suggestion would be to go back and clarify the goal of the DR and reformulate it more succinctly.

The issue is pretty clear: add timestamp to entities, and that's what this DR is about, in the last PR there were doubt about the extraction of the `Entity` superclass, so it was removed, in my opinion it's just a refactoring detail, can be extracted or not.
If I'm not wrong no DR were made for the addition of the `createdTimestamp` to `TransferProcess`, probably that was the big mistake.
However I would say that this modification is pretty simple and straightforward, the DR makes completely sense, for further refactoring (e.g. extraction of an `Entity` superclass) let's discuss later on

**Pull request closed without merging**